### PR TITLE
Remove auto-decoding from startsWith when Needles are all narrow chars

### PR DESCRIPTION
**Pull request by JackStouffer**

### CODE REVIEW

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -4104,7 +4104,22 @@ if (isInputRange!Range && Needles.length > 1 &&
     is(typeof(.startsWith!pred(doesThisStart, withOneOfThese[0])) : bool ) &&
     is(typeof(.startsWith!pred(doesThisStart, withOneOfThese[1 .. $])) : uint))
 {
-    alias haystack = doesThisStart;
+    import std.meta : allSatisfy;
+
+    template checkType(T)
+    {
+        enum checkType = is(Unqual!(ElementEncodingType!Range) == Unqual!T);
```
Comment (BBasile) [inline std/algorithm/searching.d]: Any particular reason why you use a template and not a templated enum ?

**Comment by n8sh** [comment]
You need to check that that `startsWith` is using the default predicate `"a == b"` for this to be sound.

**Comment by wilzbach** [comment]
> You need to check that that startsWith is using the default predicate "a == b" for this to be sound.

Lambda comparison has been merged a few days ago:

https://dlang.org/changelog/pending.html#lambdacomp

**Comment by schveiguy** [comment]
> Lambda comparison has been merged a few days ago:

There's a list issue with that, but I don't know where the source is, so not sure how to change it.

It looks like item 12 really should be a sub-bullet of item 11.

**Comment by wilzbach** [comment]
>  but I don't know where the source is, so not sure how to change it.

https://github.com/dlang/dmd/tree/master/changelog

I will add a few notes and links in the header.

**Comment by schveiguy** [comment]
Hm... the docs look correct to me (The parentheses match as expected), so maybe it's an issue with the thing that assembles it together? Where is that code?

Sorry for hijacking this thread, don't know of a better place to ask these questions...

**Comment by wilzbach** [comment]
> Hm... the docs look correct to me (The parentheses match as expected), so maybe it's an issue with the thing that assembles it together? Where is that code?

https://github.com/dlang/tools/blob/master/changed.d
https://github.com/dlang/dlang.org/blob/master/posix.mak#L900

> Sorry for hijacking this thread, don't know of a better place to ask these questions...

Slack?

**Comment by wilzbach** [comment]
@schveiguy fix for the changelog entry: https://github.com/dlang/dmd/pull/7847

**Comment by JackStouffer** [comment]
What does `isSame` return if it can't compare the lambdas? Does anyone know?

**Comment by wilzbach** [comment]
`false`

File: std/algorithm/searching.d [hunk 2]
```diff
@@ -4104,7 +4104,24 @@ if (isInputRange!Range && Needles.length > 1 &&
...
+    }
+
+    // auto-decoding special case
+    static if (__traits(isSame, binaryFun!pred, binaryFun!"a == b") &&
```
Comment (wilzbach) [inline std/algorithm/searching.d]: Use `alias pred = "a == b"` to `(a, b) => a == b` and then you can use here `(a, b) => a == b`
No need to stress the compiler more than necessary with mixins.

**Comment by andralex** [comment]
Cool! I think things are fine as they are to avoid too much change of interface.

**Comment by andralex** [comment]
Does this help with using `skipOver` to skip over the BOM per discussion in https://github.com/dlang/phobos/pull/6114? If so, adding some examples with this PR would be awesome.

**Comment by wilzbach** [comment]
> Cool! I think things are fine as they are to avoid too much change of interface.

FWIW my comment was just about reducing the compile-time by avoiding an unnecessary mixin (and the longer-term goal to fade out string functions). I don't think that the interface of a template would be affected by changing by changing the default value of a parameter.

**Comment by JackStouffer** [comment]
>Does this help with using skipOver to skip over the BOM per discussion in #6114?

probably not, as `\uFEFF` is always going to be a `dchar` right?

**Pull request merged**